src/conversions/chrono.rs

```diff
@@ -0,0 +1,573 @@
+#![cfg(feature = "chrono")]
+
+//! Conversions to and from [chrono](https://docs.rs/chrono/)’s `Duration`,
+//! `NaiveDate`, `NaiveTime`, `DateTime<Tz>`, `FixedOffset`, and `Utc`.
+//!
+//! # Setup
+//!
+//! To use this feature, add this to your **`Cargo.toml`**:
+//!
+//! ```toml
+//! [dependencies]
+//! # change * to the latest versions
+//! pyo3 = { version = "*", features = ["chrono"] }
+//! chrono = "0.4"
+// workaround for `extended_key_value_attributes`: https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643
+#![cfg_attr(docsrs, cfg_attr(docsrs, doc = concat!("pyo3 = { version = \"", env!("CARGO_PKG_VERSION"),  "\", features = [\"chrono\"] }")))]
+#![cfg_attr(
+    not(docsrs),
+    doc = "pyo3 = { version = \"*\", features = [\"chrono\"] }"
+)]
+//! ```
+//!
+//! Note that you must use compatible versions of chrono and PyO3.
+//! The required chrono version may vary based on the version of PyO3.
+use crate::exceptions::PyTypeError;
+use crate::types::{
+    timezone_utc, PyDate, PyDateAccess, PyDateTime, PyDelta, PyDeltaAccess, PyTime, PyTimeAccess,
+    PyTzInfo, PyTzInfoAccess,
+};
+use crate::{FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python, ToPyObject};
+use chrono::offset::{FixedOffset, Utc};
+use chrono::{
+    DateTime, Datelike, Duration, NaiveDate, NaiveDateTime, NaiveTime, Offset, TimeZone, Timelike,
+};
+use std::convert::TryInto;
+
+impl ToPyObject for Duration {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        // Total number of days
+        let days = self.num_days();
+        // Remainder of seconds
+        let secs_dur = *self - Duration::days(days);
+        // .try_into() converts i64 to i32, but this should never overflow
+        // since it's at most the number of seconds per day
+        let secs = secs_dur.num_seconds().try_into().unwrap();
+        // Fractional part of the microseconds
+        let micros = (secs_dur - Duration::seconds(secs_dur.num_seconds()))
+            .num_microseconds()
+            // This should never panic since we are just getting the fractional
+            // part of the total microseconds, which should never overflow.
+            .unwrap()
+            // Same for the conversion from i64 to i32
+            .try_into()
+            .unwrap();
+
+        // We do not need to check i64 to i32 cast from rust because
+        // python will panic with OverflowError.
+        // Not sure if we need normalize here.
+        let delta = PyDelta::new(py, days.try_into().unwrap_or(i32::MAX), secs, micros, false)
+            .expect("Failed to construct delta");
```

Comment:
I wasn't sure what to do here.

Reply:
I'll need to look at the implementation because this is something that is not present in the python api, I'll comment once I know more

Reply:
Ok, I found the function that is called in cpython: https://github.com/python/cpython/blob/023c51d9d8e2fd91069eea2bf12e373f1c71e9d2/Modules/_datetimemodule.c#L1109
```
Normalize the members iff normalize is
true.  Passing false is a speed optimization, if you know for sure
that seconds and microseconds are already in their proper ranges.  In any
case, raises OverflowError and returns NULL if the normalized days is out
of range.
```

Since we don't check that the data is in proper range, we should probably pass `true` here

Reply:
Yes, but I am not sure if we need to normalize, if I am not wrong, it converts the microseconds to seconds if it is like too much.

Reply:
So, I did some more invetigation, and not normalizing the pydelta might lead to some problems, and we should do more checks/normalization ourselves if we don't want to normalize, so I'm going to set it to true.
For example this test won't pass if we don't normalize:
```rust
let delta = Duration::days(-1) + Duration::seconds(1) + Duration::microseconds(-10);
Python::with_gil(|py| {
    let delta = delta.to_object(py);
    let delta: &PyDelta = delta.extract(py).unwrap();
    let py_delta = PyDelta::new(py, -1, 1, -10, true).unwrap();
    assert!(delta.eq(py_delta).unwrap());
});
```

But if I compare the result of calling `.total_seconds()` they return the same result